### PR TITLE
Support temporary login sessions

### PR DESCRIPTION
**Pull request by parisholley**
Allow web application to change the authorization per request instead of modifying session for active user

---
This change is [Reviewable](https://reviewable.io/reviews/apinstein/phocoa/58)

### CODE REVIEW

File: phocoa/framework/WFAuthorization.php [hunk 1]
```diff
@@ -535,11 +535,15 @@ function login($username, $password, $passIsToken = false)
...
-    function loginAsAuthorizationInfo($authInfo, $authorizeRecentLogin = true)
+    function loginAsAuthorizationInfo($authInfo, $authorizeRecentLogin = true, $temporary = false)
     {
         if (!$authInfo instanceof WFAuthorizationInfo) throw new WFException("WFAuthorizationInfo or subclass required.");
 
         $this->authorizationInfo = $authInfo;
+        
+        if($temporary){
+      			return;
+      		}
```
Comment (apinstein) [inline phocoa/framework/WFAuthorization.php]: @parisholley not sure what to make of this... does this still need reviewing? w/o some docs/context I am not sure how to eval.

Comment (parisholley) [inline phocoa/framework/WFAuthorization.php]: in my implementation, i just fetch the user out of the db based on the request header and set this for the request only. by adding "temporary" it prevents writing into the session and influencing other requests however everything that is using the shared manager static method can see this "temporary" user until the request is over.

**Comment by apinstein** [comment]
Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[phocoa/framework/WFAuthorization.php, line 546 at r1](https://reviewable.io:443/reviews/apinstein/phocoa/58#-KSCQsWVpfM9lIibwA6D-r1-546:-KSElk3i91ZO0kizJbFU:b-qq8lhk) ([raw file](https://github.com/apinstein/phocoa/blob/0061067612cf35d45945bb33c5e035cc3e058b60/phocoa/framework/WFAuthorization.php#L546)):*
<details><summary><i>Previously, parisholley (Paris Holley) wrote…</i></summary>
> in my implementation, i just fetch the user out of the db based on the request header and set this for the request only. by adding "temporary" it prevents writing into the session and influencing other requests however everything that is using the shared manager static method can see this "temporary" user until the request is over.
</details>
Maybe it'd make more sense if I saw the other side of the code, but feels pretty hacky. Are you calling I am having trouble reasoning about what might break b/c of where you patched in.

This func `loginAsAuthorizationInfo` that you patched, it is interestingly only used by a few tests.  Not actual code. It's only use in non-test code seems to be when it's used by the `WFAuthorizationManager` class internally.

After looking at it more now I realized I didn't notice before that you allow the "instance" var to be set but avoid persisting stuff to the session. Presumably you're just calling `$mgr->loginAsAuthorizationInfo(.., .., true)` with the data from the magic header to implement this?

I think I see how you got it to work.

Here's my areas of concern: 
- there are places where the auth mgr references the session directly, such as in `isLoggedIn()` and `recentLoginTime`, and having that data out-of-sync w/the `authInfo` seems risky. Despite the `loginAsAuthorizationInfo()`  docs that indicate this is a great place to do this, 

```
     * Change the current session authorization info to the specified WFAuthorizationInfo object.
     * This is useful for SSO or other internal user-switching capabilities.
```

That func also ensures that the session data matches the authInfo data. That's the risky part for me. I am not sure the subtle ways it could break.

I feel like a safer solution might be to alter the auth mgr class to have a delegate for `authorizationInfo()` and just implement your "temp hack" in userland? 

So the only change to phocoa is to `WFAuthorization::authorizationInfo()`, to leverage a delegate func of same name (add it to `WFAuthorizationDelegate`), 

```
    function authorizationInfo()
    {
        // give delegate a chance to munge current auth-info
        if (method_exists($this->authorizationDelegate, 'loginInvocationPath'))
        {
            return $this->authorizationDelegate->authorizationInfo($this->authorizationInfo);
        }
        else
        {
            return $this->authorizationInfo;
        }
    }```

and then in userland you'd just munge it there. Then, there are no weird race conditions with how the temp authinfo gets set, or places where the session / instance are out-of-sync, etc.

Maybe a little pedantic, but I think it's easier to reason about and seems less risky from security perspective. Seems less side-effecty than the current implementation, and also has the effect of consolidating the "check header" logic and "munge auth info" logic into the same spot.

Thoughts?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/apinstein/phocoa/58)*
<!-- Sent from Reviewable.io -->

**Comment by parisholley** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[phocoa/framework/WFAuthorization.php, line 546 at r1](https://reviewable.io:443/reviews/apinstein/phocoa/58#-KSCQsWVpfM9lIibwA6D-r1-546:-KSEqXmt5s-YTU1-obA8:b-n7wg0n) ([raw file](https://github.com/apinstein/phocoa/blob/0061067612cf35d45945bb33c5e035cc3e058b60/phocoa/framework/WFAuthorization.php#L546)):*
<details><summary><i>Previously, apinstein (Alan Pinstein) wrote…</i></summary>
> Maybe it'd make more sense if I saw the other side of the code, but feels pretty hacky. Are you calling I am having trouble reasoning about what might break b/c of where you patched in.
> 
> This func `loginAsAuthorizationInfo` that you patched, it is interestingly only used by a few tests.  Not actual code. It's only use in non-test code seems to be when it's used by the `WFAuthorizationManager` class internally.
> 
> After looking at it more now I realized I didn't notice before that you allow the "instance" var to be set but avoid persisting stuff to the session. Presumably you're just calling `$mgr->loginAsAuthorizationInfo(.., .., true)` with the data from the magic header to implement this?
> 
> I think I see how you got it to work.
> 
> Here's my areas of concern: 
> - there are places where the auth mgr references the session directly, such as in `isLoggedIn()` and `recentLoginTime`, and having that data out-of-sync w/the `authInfo` seems risky. Despite the `loginAsAuthorizationInfo()`  docs that indicate this is a great place to do this, 
> 
> ```
>      * Change the current session authorization info to the specified WFAuthorizationInfo object.
>      * This is useful for SSO or other internal user-switching capabilities.
> ```
> 
> That func also ensures that the session data matches the authInfo data. That's the risky part for me. I am not sure the subtle ways it could break.
> 
> I feel like a safer solution might be to alter the auth mgr class to have a delegate for `authorizationInfo()` and just implement your "temp hack" in userland? 
> 
> So the only change to phocoa is to `WFAuthorization::authorizationInfo()`, to leverage a delegate func of same name (add it to `WFAuthorizationDelegate`), 
> 
> ```
>     function authorizationInfo()
>     {
>         // give delegate a chance to munge current auth-info
>         if (method_exists($this->authorizationDelegate, 'loginInvocationPath'))
>         {
>             return $this->authorizationDelegate->authorizationInfo($this->authorizationInfo);
>         }
>         else
>         {
>             return $this->authorizationInfo;
>         }
>     }```
> 
> and then in userland you'd just munge it there. Then, there are no weird race conditions with how the temp authinfo gets set, or places where the session / instance are out-of-sync, etc.
> 
> Maybe a little pedantic, but I think it's easier to reason about and seems less risky from security perspective. Seems less side-effecty than the current implementation, and also has the effect of consolidating the "check header" logic and "munge auth info" logic into the same spot.
> 
> Thoughts?
> 
</details>
Correct, I went this route as the user land implementation you mentioned didn't exist (but seems more appropriate). There is already a convention of going to "user land"  to delegate the login logic and many auth frameworks also have the same process for delegating how the session is serialized/unserialized to produce a "user object".

That being said, I'm not sure the delegate will solve any sync/out of sync issue. This call was being made right after manager is initialized, and you will still have the potential risk of logic somewhere depending on the session/auth object being in sync (but in theory, no one should be inspecting the session directly).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/apinstein/phocoa/58)*

**Comment by apinstein** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[phocoa/framework/WFAuthorization.php, line 538 at r1](https://reviewable.io:443/reviews/apinstein/phocoa/58#-KSEuMkZW7wyKbTL3vGT:-KSEuMkZW7wyKbTL3vGU:b-k2ywt9) ([raw file](https://github.com/apinstein/phocoa/blob/0061067612cf35d45945bb33c5e035cc3e058b60/phocoa/framework/WFAuthorization.php#L538)):*
> ```PHP
>      * @param object WFAuthorizationInfo The new authorization info to set for the current session.
>      */
>     function loginAsAuthorizationInfo($authInfo, $authorizeRecentLogin = true, $temporary = false)
> ```

if you do keep this, doc the param please.

---

*[phocoa/framework/WFAuthorization.php, line 546 at r1](https://reviewable.io:443/reviews/apinstein/phocoa/58#-KSCQsWVpfM9lIibwA6D-r1-546:-KSEstjFeFTxYfbDEm_f:b-248dqw) ([raw file](https://github.com/apinstein/phocoa/blob/0061067612cf35d45945bb33c5e035cc3e058b60/phocoa/framework/WFAuthorization.php#L546)):*
<details><summary><i>Previously, parisholley (Paris Holley) wrote…</i></summary>
> Correct, I went this route as the user land implementation you mentioned didn't exist (but seems more appropriate). There is already a convention of going to "user land"  to delegate the login logic and many auth frameworks also have the same process for delegating how the session is serialized/unserialized to produce a "user object".
> 
> That being said, I'm not sure the delegate will solve any sync/out of sync issue. This call was being made right after manager is initialized, and you will still have the potential risk of logic somewhere depending on the session/auth object being in sync (but in theory, no one should be inspecting the session directly).
</details>
I think you're right in terms of it doesn't change either the fact that there is data out-of-sync, or what happens if somone reaches into the sessino (tho they deserve it in that case :)

I think the only argument to move it to delegate is just it's slightly more readable/grokkable/consistent, which I think you agree with.

I am comfortable with whichever way you choose. Thanks for the discussion.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/apinstein/phocoa/58)*

**Comment by parisholley** [comment]
Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions.

---

*[phocoa/framework/WFAuthorization.php, line 538 at r1](https://reviewable.io:443/reviews/apinstein/phocoa/58#-KSEuMkZW7wyKbTL3vGT:-KSHWLQU53_93xWoC20P:b76fhpc) ([raw file](https://github.com/apinstein/phocoa/blob/0061067612cf35d45945bb33c5e035cc3e058b60/phocoa/framework/WFAuthorization.php#L538)):*
<details><summary><i>Previously, apinstein (Alan Pinstein) wrote…</i></summary>
> if you do keep this, doc the param please.
</details>
implemented delegate approach

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/apinstein/phocoa/58)*

**Comment by apinstein** [comment]
:lgtm:

---

Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/apinstein/phocoa/58#-:-KSYlAmcSt0mjT7P14WD:bnfp4nl)*

**Pull request merged**